In [1]:
# Basic Import
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns

# Classification models
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import (
    RandomForestClassifier,
    AdaBoostClassifier,
    GradientBoostingClassifier
)
from sklearn.svm import SVC
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
    classification_report
)

In [3]:
import sys

!{sys.executable} -m pip install numpy pandas matplotlib seaborn scikit-learn scipy statsmodels xgboost catboost

  Using cached matplotlib-3.11.0-cp311-cp311-win_amd64.whl.metadata (80 kB)
  Using cached seaborn-0.13.2-py3-none-any.whl.metadata (5.4 kB)
  Using cached scipy-1.17.1-cp311-cp311-win_amd64.whl.metadata (60 kB)
  Using cached statsmodels-0.14.6-cp311-cp311-win_amd64.whl.metadata (9.8 kB)
  Using cached contourpy-1.3.3-cp311-cp311-win_amd64.whl.metadata (5.5 kB)
  Using cached cycler-0.12.1-py3-none-any.whl.metadata (3.8 kB)
  Using cached fonttools-4.63.0-cp311-cp311-win_amd64.whl.metadata (121 kB)
  Using cached kiwisolver-1.5.0-cp311-cp311-win_amd64.whl.metadata (5.2 kB)
  Using cached pillow-12.2.0-cp311-cp311-win_amd64.whl.metadata (9.0 kB)
  Using cached pyparsing-3.3.2-py3-none-any.whl.metadata (5.8 kB)
  Using cached joblib-1.5.3-py3-none-any.whl.metadata (5.5 kB)
  Using cached threadpoolctl-3.6.0-py3-none-any.whl.metadata (13 kB)
  Using cached patsy-1.0.2-py2.py3-none-any.whl.metadata (3.6 kB)
  Using cached plotly-6.8.0-py3-none-any.whl.metadata (9.0 kB)
Using cached matplo

  You can safely remove it manually.

[notice] A new release of pip is available: 24.0 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import sys

!{sys.executable} -m pip install --upgrade --force-reinstall numpy pandas

  Using cached numpy-2.4.6-cp311-cp311-win_amd64.whl.metadata (6.6 kB)
  Using cached python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
  Using cached tzdata-2026.2-py2.py3-none-any.whl.metadata (1.4 kB)
  Using cached six-1.17.0-py2.py3-none-any.whl.metadata (1.7 kB)
Using cached numpy-2.4.6-cp311-cp311-win_amd64.whl (12.6 MB)
   ---------------------------------------- 0.0/9.9 MB ? eta -:--:--
   ---------------------------------------- 0.0/9.9 MB ? eta -:--:--
   - -------------------------------------- 0.3/9.9 MB 5.2 MB/s eta 0:00:02
   ----- ---------------------------------- 1.3/9.9 MB 11.6 MB/s eta 0:00:01
   ---------- ----------------------------- 2.5/9.9 MB 17.7 MB/s eta 0:00:01
   -------------- ------------------------- 3.5/9.9 MB 18.7 MB/s eta 0:00:01
   ------------------- -------------------- 4.8/9.9 MB 20.4 MB/s eta 0:00:01
   ----------------------- ---------------- 5.9/9.9 MB 20.9 MB/s eta 0:00:01
   ---------------------------- ----------- 7.1/9.9 MB

  You can safely remove it manually.
  You can safely remove it manually.
  You can safely remove it manually.

[notice] A new release of pip is available: 24.0 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


# Importing Raw Data

In [147]:
import pandas as pd
df=pd.read_csv('data/raw.csv')

In [148]:
df.shape

(1000, 21)

Show Top 5 Records

In [149]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,1024-GUALD,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,24.80,24.8,Yes
1,0484-JPBRU,Male,0,No,No,41,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Bank transfer (automatic),25.25,996.45,No
2,3620-EHIMZ,Female,0,Yes,Yes,52,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.35,1031.7,No
3,6910-HADCM,Female,0,No,No,1,Yes,No,Fiber optic,No,...,Yes,No,No,No,Month-to-month,No,Electronic check,76.35,76.35,Yes
4,8587-XYZSF,Male,0,No,No,67,Yes,No,DSL,No,...,No,Yes,No,No,Two year,No,Bank transfer (automatic),50.55,3260.1,No


In [150]:
df["TotalCharges"] = pd.to_numeric(
    df["TotalCharges"],
    errors="coerce"
)

In [151]:
df["TotalCharges"].isnull().sum()

np.int64(1)

In [152]:
# Filling missing values

In [153]:
df["TotalCharges"] = df["TotalCharges"].fillna(
    df["TotalCharges"].median()
)

In [154]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

# Finding LOG Transformation needed or not

In [155]:
num_cols = df.select_dtypes(include=np.number).columns
cat_cols = df.select_dtypes(include='object').columns

C:\Users\kotha\AppData\Local\Temp\ipykernel_20196\228636448.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df.select_dtypes(include='object').columns


In [156]:
num_cols

Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges'], dtype='str')

In [157]:
df[num_cols].skew()

SeniorCitizen     1.817592
tenure            0.261901
MonthlyCharges   -0.217870
TotalCharges      1.031874
dtype: float64

Interpretation:

- skewness between -0.5 to 0.5 → approximately normal
- 0.5 to 1 → moderate skew
- >1 → highly skewed


| Column         | Skewness | Meaning                              | Log Transformation? |
| -------------- | -------: | ------------------------------------ | ------------------- |
| SeniorCitizen  |     1.81 | Highly right skewed                  | ❌ No                |
| tenure         |     0.26 | Almost normal                        | ❌ No                |
| MonthlyCharges |    -0.21 | Almost normal (slightly left skewed) | ❌ No                |
| TotalCharges   |     1.03 | Right skewed                         | ✅ Yes               |


**Important point**: SeniorCitizen should not get log transformation even though skewness is high.
- Because SeniorCitizen is a binary column:
   - 0 = No
   - 1 = Yes



# Applying Log Transformation

In [158]:
df["TotalCharges"] = np.log1p(
    df["TotalCharges"]
)

In [159]:
# Checking now skew is there or not
df["TotalCharges"].skew()

np.float64(-0.7524672881284353)

# Split features and target

In [160]:
# Split features and target
X = df.drop('Churn', axis=1)
y = df['Churn']

In [161]:
X.shape

(1000, 20)

In [162]:
y.shape

(1000,)

# Separating numerical features and catogerical features

In [163]:

num_cols = df.select_dtypes(include=np.number).columns
cat_cols = df.select_dtypes(include='object').columns

C:\Users\kotha\AppData\Local\Temp\ipykernel_20196\424073305.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df.select_dtypes(include='object').columns


In [164]:
print(num_cols)

Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges'], dtype='str')


In [165]:
print(cat_cols)

Index(['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService',
       'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup',
       'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies',
       'Contract', 'PaperlessBilling', 'PaymentMethod', 'Churn'],
      dtype='str')


# Checking the how many categorical there per feature

Finding which is better LabelEncoder or TF-IDF Vectorizer to apply on categorical columns

In [166]:
category_columns = []
text_columns = []

for col in X.select_dtypes(include='object').columns:
    
    # average word count
    avg_words = X[col].astype(str).apply(
        lambda x: len(x.split())
    ).mean()
    
    print(col, "Average Words:", avg_words)
    
    # decision
    if avg_words <= 2:
        category_columns.append(col)
    else:
        text_columns.append(col)

print("Category Columns:", category_columns)

print("Text Columns:", text_columns)

customerID Average Words: 1.0
gender Average Words: 1.0
Partner Average Words: 1.0
Dependents Average Words: 1.0
PhoneService Average Words: 1.0
MultipleLines Average Words: 1.178
InternetService Average Words: 1.436
OnlineSecurity Average Words: 1.45
OnlineBackup Average Words: 1.45
DeviceProtection Average Words: 1.45
TechSupport Average Words: 1.45
StreamingTV Average Words: 1.45
StreamingMovies Average Words: 1.45
Contract Average Words: 1.431
PaperlessBilling Average Words: 1.0
PaymentMethod Average Words: 2.435
Category Columns: ['customerID', 'gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling']
Text Columns: ['PaymentMethod']


C:\Users\kotha\AppData\Local\Temp\ipykernel_20196\802263798.py:4: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in X.select_dtypes(include='object').columns:


Observation: we can apply the OneHotEncoder because we almost one type word only.

In [167]:

num_cols = X.select_dtypes(include=np.number).columns
cat_cols = X.select_dtypes(include='object').columns

C:\Users\kotha\AppData\Local\Temp\ipykernel_20196\1932948254.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = X.select_dtypes(include='object').columns


In [168]:
num_cols

Index(['SeniorCitizen', 'tenure', 'MonthlyCharges', 'TotalCharges'], dtype='str')

In [171]:
# Create Column Transformer with 3 types of transformers
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer

numeric_transformer = StandardScaler()
oh_transformer = OneHotEncoder()

preprocessor = ColumnTransformer(
    [
        ("OneHotEncoder", oh_transformer, cat_cols),
         ("StandardScaler", numeric_transformer, num_cols),        
    ]
)

In [172]:
X = preprocessor.fit_transform(X)

In [173]:
X

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 20000 stored elements and shape (1000, 1045)>

In [174]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

y = le.fit_transform(y)

In [175]:
np.unique(y)

array([0, 1])

observation: now see each row converted to the separate binary column

# separate dataset into train and test

In [176]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)
X_train.shape, X_test.shape

((800, 1045), (200, 1045))

# Evaluation function

In [177]:
def evaluate_model(true, predicted, probability):

    accuracy = accuracy_score(true,predicted)
    precision = precision_score(true,predicted)
    recall = recall_score(true,predicted)
    f1 = f1_score(true,predicted)
    roc_auc = roc_auc_score(true,probability )

    return (
        accuracy,
        precision,
        recall,
        f1,
        roc_auc
    )

# Models without hyperparameter tuning

In [178]:
models = {

"Logistic Regression":LogisticRegression(max_iter=1000),
"KNN":KNeighborsClassifier(),
"Decision Tree":DecisionTreeClassifier(),
"Random Forest":RandomForestClassifier(),
"AdaBoost":AdaBoostClassifier(),
"Gradient Boost":GradientBoostingClassifier(),
"XGBoost":XGBClassifier(),
"CatBoost":CatBoostClassifier(verbose=False),
"SVM":SVC(probability=True)

}

# Train models without tuning

In [179]:
import numpy as np

np.isnan(X_train.toarray()).sum()

np.int64(0)

In [180]:
import numpy as np

# find location of NaN
nan_position = np.argwhere(np.isnan(X.toarray()))

nan_position

array([], shape=(0, 2), dtype=int64)

In [181]:
model_result = []


for name, model in models.items():
    model.fit( X_train, y_train)
    y_pred = model.predict(X_test)
    y_prob = model.predict_proba( X_test)[:,1]
    accuracy, precision, recall, f1, roc_auc = evaluate_model(
        y_test,
        y_pred,
        y_prob
    )
    print("="*50)
    print(name)

    print("Accuracy:",accuracy)
    print( "Precision:", precision)
    print( "Recall:",recall)
    print("F1:",f1)
    print("ROC AUC:",roc_auc)


    model_result.append(
        [
            name,
            accuracy,
            precision,
            recall,
            f1,
            roc_auc
        ]
    )

Logistic Regression
Accuracy: 0.84
Precision: 0.7105263157894737
Recall: 0.5625
F1: 0.627906976744186
ROC AUC: 0.8433388157894737
KNN
Accuracy: 0.775
Precision: 0.5319148936170213
Recall: 0.5208333333333334
F1: 0.5263157894736842
ROC AUC: 0.8026315789473684
Decision Tree
Accuracy: 0.805
Precision: 0.6216216216216216
Recall: 0.4791666666666667
F1: 0.5411764705882353
ROC AUC: 0.693530701754386
Random Forest
Accuracy: 0.795
Precision: 0.5945945945945946
Recall: 0.4583333333333333
F1: 0.5176470588235295
ROC AUC: 0.8364172149122807
AdaBoost
Accuracy: 0.81
Precision: 0.65625
Recall: 0.4375
F1: 0.525
ROC AUC: 0.8447779605263158
Gradient Boost
Accuracy: 0.83
Precision: 0.6842105263157895
Recall: 0.5416666666666666
F1: 0.6046511627906976
ROC AUC: 0.8634868421052632
XGBoost
Accuracy: 0.79
Precision: 0.575
Recall: 0.4791666666666667
F1: 0.5227272727272727
ROC AUC: 0.819078947368421
CatBoost
Accuracy: 0.81
Precision: 0.625
Recall: 0.5208333333333334
F1: 0.5681818181818182
ROC AUC: 0.85279605263157

c:\Users\kotha\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\svm\_base.py:239: FutureWarning: The `probability` parameter was deprecated in 1.9 and will be removed in version 1.11. Use `CalibratedClassifierCV(SVC(), ensemble=False)` instead of `SVC(probability=True)`
  warnings.warn(


SVM
Accuracy: 0.82
Precision: 0.6666666666666666
Recall: 0.5
F1: 0.5714285714285714
ROC AUC: 0.8399122807017544


In [182]:
result_df = pd.DataFrame(
    model_result,
    columns=[
        "Model",
        "Accuracy",
        "Precision",
        "Recall",
        "F1",
        "ROC_AUC"
    ]
)


result_df.sort_values(
    by="ROC_AUC",
    ascending=False
)

,Model,Accuracy,Precision,Recall,F1,ROC_AUC
5,Gradient Boost,0.830,0.684211,0.541667,0.604651,0.863487
7,CatBoost,0.810,0.625000,0.520833,0.568182,0.852796
4,AdaBoost,0.810,0.656250,0.437500,0.525000,0.844778
0,Logistic Regression,0.840,0.710526,0.562500,0.627907,0.843339
8,SVM,0.820,0.666667,0.500000,0.571429,0.839912
3,Random Forest,0.795,0.594595,0.458333,0.517647,0.836417
6,XGBoost,0.790,0.575000,0.479167,0.522727,0.819079
1,KNN,0.775,0.531915,0.520833,0.526316,0.802632
2,Decision Tree,0.805,0.621622,0.479167,0.541176,0.693531


# Hyperparameter tuning with Cross Validation

In [140]:
from sklearn.model_selection import GridSearchCV

In [183]:
models_params = {


"Logistic Regression":
(
    LogisticRegression(max_iter=1000),
    {
        "C":[0.01,0.1,1,10],
        "solver":["liblinear","lbfgs"]
    }
),

"KNN":
(
    KNeighborsClassifier(),
    {
        "n_neighbors":[3,5,7,9],
        "weights":["uniform","distance"]
    }
),

"Decision Tree":
(
    DecisionTreeClassifier(),

    {
        "max_depth":[5,10,15,None],
        "min_samples_split":[2,5,10],
        "criterion":["gini","entropy"]
    }
),

"Random Forest":
(
    RandomForestClassifier(),

    {
        "n_estimators":[100,200,300],
        "max_depth":[5,10,20,None],
        "min_samples_split":[2,5,10]
    }
),

"AdaBoost":
(
    AdaBoostClassifier(),

    {
        "n_estimators":[50,100,200],
        "learning_rate":[0.01,0.1,1]
    }
),

"Gradient Boost":

(
    GradientBoostingClassifier(),

    {
        "n_estimators":[100,200],
        "learning_rate":[0.01,0.1],
        "max_depth":[3,5,7]
    }
),

"XGBoost":
(
    XGBClassifier(),
    {
        "n_estimators":[100,200],
        "max_depth":[3,5,7],
        "learning_rate":[0.01,0.1]
    }
),

"CatBoost":
(
    CatBoostClassifier(verbose=False),
    {
        "iterations":[100,200],
        "depth":[4,6,8],
        "learning_rate":[0.01,0.1]
    }
)

}

# Evaluate tuned model

In [184]:
tuned_results = []


for name,(model,params) in models_params.items():

    print("="*50)

    print("Training:",name)


    grid = GridSearchCV(
        estimator=model,
        param_grid=params,
        cv=5,
        scoring="roc_auc",
        n_jobs=-1
    )


    grid.fit(
        X_train,
        y_train
    )


    best_model = grid.best_estimator_


    y_pred = best_model.predict(
        X_test
    )


    y_prob = best_model.predict_proba(
        X_test
    )[:,1]



    accuracy,precision,recall,f1,roc_auc = evaluate_model(
        y_test,
        y_pred,
        y_prob
    )


    print("Best Parameters:")
    print(grid.best_params_)


    print("ROC AUC:",roc_auc)


    tuned_results.append(
        [
            name,
            accuracy,
            precision,
            recall,
            f1,
            roc_auc
        ]
    )

Training: Logistic Regression
Best Parameters:
{'C': 0.01, 'solver': 'lbfgs'}
ROC AUC: 0.8543037280701754
Training: KNN
Best Parameters:
{'n_neighbors': 9, 'weights': 'distance'}
ROC AUC: 0.813733552631579
Training: Decision Tree
Best Parameters:
{'criterion': 'entropy', 'max_depth': 5, 'min_samples_split': 10}
ROC AUC: 0.7890625
Training: Random Forest
Best Parameters:
{'max_depth': 20, 'min_samples_split': 10, 'n_estimators': 300}
ROC AUC: 0.8600603070175438
Training: AdaBoost
Best Parameters:
{'learning_rate': 0.1, 'n_estimators': 200}
ROC AUC: 0.8660910087719298
Training: Gradient Boost
Best Parameters:
{'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200}
ROC AUC: 0.8652686403508771
Training: XGBoost
Best Parameters:
{'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 100}
ROC AUC: 0.8386787280701755
Training: CatBoost
Best Parameters:
{'depth': 4, 'iterations': 200, 'learning_rate': 0.1}
ROC AUC: 0.8660910087719298


# Convert tunned results to dataframe

In [185]:
tuned_df = pd.DataFrame(

    tuned_results,

    columns=[
        "Model",
        "Accuracy",
        "Precision",
        "Recall",
        "F1",
        "ROC_AUC"
    ]

)


tuned_df.sort_values(
    by="ROC_AUC",
    ascending=False
)

,Model,Accuracy,Precision,Recall,F1,ROC_AUC
7,CatBoost,0.850,0.750000,0.562500,0.642857,0.866091
4,AdaBoost,0.830,0.718750,0.479167,0.575000,0.866091
5,Gradient Boost,0.840,0.722222,0.541667,0.619048,0.865269
3,Random Forest,0.820,0.700000,0.437500,0.538462,0.860060
0,Logistic Regression,0.830,0.791667,0.395833,0.527778,0.854304
6,XGBoost,0.825,0.658537,0.562500,0.606742,0.838679
1,KNN,0.790,0.565217,0.541667,0.553191,0.813734
2,Decision Tree,0.795,0.574468,0.562500,0.568421,0.789062


In [ ]:
Gradient Boost	0.830	0.684211	0.541667	0.604651	0.863487
CatBoost	0.810	0.625000	0.520833	0.568182	0.852796

In [ ]:
Data Loading
      ↓
Data Cleaning
      ↓
EDA
      ↓
Check skewness of numerical columns
      ↓
Apply log transformation (only skewed columns)
      ↓
Preprocessing (Scaling + Encoding)
      ↓
Train models
      ↓
Hyperparameter tuning

In [ ]:
Data Loading
      ↓
Data Cleaning(need to fill the totalcharges with median)
      ↓
Apply log transformation (we have to apply log transformation to totalcharges)
      ↓
Preprocessing (Scaling + Encoding)
      ↓
we have to apply the labelencoding to y as well
       ↓
Hyperparameter tuning
      ↓
Train models
      ↓
now evalution

this what we have to do in .py file ok remember this why like this only becasue we tried to check
the in juptyernotebook what filling need apply transform or not preprocessing for x and y and hyper because
got better accuracy then evualtion tell me if i miss anything